shimmingtoolbox/unwrap/prelude.py

```diff
@@ -0,0 +1,92 @@
+    can optionally be saved.
+
+    Args:
+        wrapped_phase (numpy.ndarray): 3D radian numpy array to perform phase unwrapping
+        mag (numpy.ndarray): 3D magnitude numpy array corresponding to the phase array
+        affine (numpy.ndarray): 2D array (4x4) containing the transformation coefficients. Can be acquired by :
+            nii = nib.load("nii_path")
+            affine = nii.affine
+        mask (numpy.ndarray, optional): numpy array of booleans with shape of `complex_array` to mask during phase unwrapping
+        path_2_unwrapped_phase (string, optional): relative or absolute path to output the nii unwrapped phase
+        is_unwrapping_in_2d (bool, optional): prelude parameter to unwrap un 2d
+        is_saving_nii (bool, optional): specify whether `complex_array`, `affine`, `mask` and `unwrapped_phase` nii files will be
+        saved
+
+    Returns:
+        numpy.ndarray: 3D array with the shape of `complex_array` of the unwrapped phase output from prelude
+    """
+    # Get absolute path
+    abs_path = os.path.abspath(path_2_unwrapped_phase)
+    data_save_directory = os.path.dirname(abs_path)
+
+    # Make sure directory exists, if not create it
+    if not os.path.exists(data_save_directory):
+        print('\nCreating directory for unwrapped phase at {}'.format(data_save_directory))
+        os.mkdir(data_save_directory)
+
+    # Make sure phase and mag are the right shape
+    if wrapped_phase.ndim != 3:
+        raise RuntimeError('wrapped_phase must be 3d')
+    if wrapped_phase.shape != mag.shape:
+        raise RuntimeError('The magnitude image (mag) must be the same shape as wrapped_phase')
+
+    # Save phase and mag images
+    phase_nii = nib.Nifti1Image(wrapped_phase, affine)
+    mag_nii = nib.Nifti1Image(mag, affine)
+    nib.save(phase_nii, os.path.join(data_save_directory, 'rawPhase.nii'))
+    nib.save(mag_nii, os.path.join(data_save_directory, 'mag.nii'))
+
+    # Fill options
+    options = ' '
+    if is_unwrapping_in_2d:
+        options = '-s '
+
+    # Add mask data and options if there is a mask provided
+    if not np.any(mask == -1):
+        # TODO: Make sure values are either 1 or 0
+        if mask.shape != wrapped_phase.shape:
+            raise RuntimeError('Mask must be the same shape as wrapped_phase')
+        mask_nii = nib.Nifti1Image(mask, affine)
+
+        options += '-m '
+        options += os.path.join(data_save_directory, 'mask.nii')
+        nib.save(mask_nii, os.path.join(data_save_directory, 'mask.nii'))
+
+    # Unwrap
+    unwrap_command = 'prelude -p {} -a {} -o {} {}'.format(os.path.join(data_save_directory, 'rawPhase'),
+                                                           os.path.join(data_save_directory, 'mag'),
+                                                           path_2_unwrapped_phase, options)
+    subprocess.run(unwrap_command, shell=True, check=True)
```

Comment:
Subprocess is normally donne so you can pass him a list of options instead of formating a string by hand:
```suggestion
    subprocess.run(["prelude","-p",os.path.join(data_save_directory, "rawPhase"), os.path.join(data_save_directory, "mag"), path_2_unwrapped_phase, "-m",os.path.join(data_save_directory, "mask.nii")], shell=True, check=True)
```

Reply:
we could also introduce the following API (for convenient printout and more intuitive cmd input):
````python
def run_subprocess(cmd):
    """
    Wrapper for subprocess.run() that enables to input cmd as a full string (easier for debugging).
    :param cmd:
    :return:
    """
    print("{}".format(cmd))
    subprocess.run(cmd.split(' '), stdout=subprocess.PIPE, text=True)
````

Reply:
of course print() would be replaced by logging in this project

Reply:
A problem occurs when a mask isn't provided and if the option "is_unwrapping_in_2d" is provided. Since both arguments are optional, then `ifs` are required and 4 different calls to prelude would be necessary to accomplish the same thing. ("options" can't be by itself because there are spaces in the string). Do you have a way to deal with that @gab-berest ?

Reply:
I guess Julien's function could help.

Reply:
>  both arguments are optional,

I would suggest tosave options in a list instead so you can add them accrodingly for their presence:
```suggestion
    options = []
    if is_unwrapping_2d:
        options.append("-s")
    ....

    subprocess.run(unwrap_command, shell=True, check=True)
```
In the end, you'll only need to pass the list of string options.

Reply:
Also, Julien's function could be a good wrapper to normalize the use of processes through the code.

Reply:
Implemented run_subprocess() function